File: src/main/java/com/rdvdev2/timetravelmod/impl/common/timemachine/TimeMachineStructure.java
package com.rdvdev2.timetravelmod.impl.common.timemachine;

import com.rdvdev2.timetravelmod.api.timemachine.ITimeMachine;
import com.rdvdev2.timetravelmod.api.timemachine.ITimeMachineUpgrade;
import com.rdvdev2.timetravelmod.api.timemachine.block.TimeMachineUpgradeBlock;
import com.rdvdev2.timetravelmod.impl.ModRegistries;
import net.minecraft.block.BlockState;
import net.minecraft.util.BlockRotation;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3i;
import net.minecraft.util.registry.Registry;
import net.minecraft.world.World;
import org.apache.commons.lang3.ArrayUtils;
import org.jetbrains.annotations.Nullable;

import java.util.*;
import java.util.stream.Collectors;

public class TimeMachineStructure {

    private static final Map<ITimeMachine, BlockState[]> UPGRADE_STATE_CACHE = new HashMap<>();

    private final BlockPos[] basicPos;
    private final BlockPos[] corePos;
    private final BlockPos[] controllerPos;
    private final BlockPos[] airPos;
    private final BlockState[] upgradeStates;
    private final ITimeMachine timeMachine;

    public TimeMachineStructure(BlockPos[] basicPos, BlockPos[] corePos, BlockPos[] controllerPos, BlockPos[] airPos, ITimeMachine timeMachine) {
        this.basicPos = basicPos;
        this.corePos = corePos;
        this.controllerPos = controllerPos;
        this.airPos = airPos;
        this.upgradeStates = generateUpgradeStates(timeMachine);
        this.timeMachine = timeMachine;
    }

    private static BlockState[] generateUpgradeStates(ITimeMachine timeMachine) {
        if (UPGRADE_STATE_CACHE.containsKey(timeMachine)) return UPGRADE_STATE_CACHE.get(timeMachine);

        List<ITimeMachineUpgrade> upgrades = ModRegistries.TIME_MACHINE_UPGRADE.stream().parallel()
                .filter(u -> u.isTimeMachineCompatible(timeMachine))
                .collect(Collectors.toList());

        UPGRADE_STATE_CACHE.put(
                timeMachine,
                Registry.BLOCK.stream().parallel()
                        .filter(b -> b instanceof TimeMachineUpgradeBlock && upgrades.contains(((TimeMachineUpgradeBlock) b).getUpgrade()))
                        .flatMap(b -> b.getStateManager().getStates().parallelStream())
                        .toArray(BlockState[]::new)
        );
        return generateUpgradeStates(timeMachine);
    }

    public BlockPos[] getBasicPos() {
        return basicPos;
    }

    public BlockPos[] getCorePos() {
        return corePos;
    }

    public BlockPos[] getControllerPos() {
        return controllerPos;
    }

    public BlockPos[] getAirPos() {
        return airPos;
    }

    public BlockState[] getControllerStates() {
        return timeMachine.getControllerStates();
    }

    public BlockState[] getCoreStates() {
        return timeMachine.getCoreStates();
    }

    public BlockState[] getBasicStates() {
        return timeMachine.getBasicStates();
    }

    @Nullable
    public BlockState[] getUpgradeStates() {
        return upgradeStates;
    }

    public ITimeMachine getTimeMachine() {
        return timeMachine;
    }

    public boolean checkIfBuilt(World world, BlockPos rootPos) {
        List<BlockState> basicStates = new ArrayList<>(Arrays.asList(getBasicStates().clone()));
        if (getUpgradeStates() != null) basicStates.addAll(Arrays.asList(getUpgradeStates().clone()));
        List<BlockState> coreStates = new ArrayList<>(Arrays.asList(getCoreStates().clone()));
        List<BlockState> controllerStates = Arrays.asList(getControllerStates().clone());
        return
                Arrays.stream(getControllerPos())
                        .map(p -> p.add(rootPos))
                        .allMatch(p -> controllerStates.contains(world.getBlockState(p)))
                && Arrays.stream(getCorePos())
                        .map(p -> p.add(rootPos))
                        .allMatch(p -> coreStates.contains(world.getBlockState(p)))
                && Arrays.stream(getAirPos())
                        .map(p -> p.add(rootPos))
                        .allMatch(world::isAir)
                && Arrays.stream(getBasicPos())
                        .map(p -> p.add(rootPos))
                        .allMatch(p -> basicStates.contains(world.getBlockState(p)));
    }

    public Map<ITimeMachineUpgrade, BlockPos> getUpgrades(World world, BlockPos rootPos) {
        if (getUpgradeStates() == null) return Collections.emptyMap();
        List<BlockState> upgradeStates = Arrays.asList(getUpgradeStates().clone());

        return Arrays.stream(getBasicPos())
                .map(p -> p.add(rootPos))
                .filter(p -> upgradeStates.contains(world.getBlockState(p)))
                .collect(Collectors.toMap(p -> ((TimeMachineUpgradeBlock) world.getBlockState(p).getBlock()).getUpgrade(), p -> p));
    }

    public BlockPos[] getStructurePos(BlockPos rootPos) {
        BlockPos[] allPos = ArrayUtils.addAll(
                ArrayUtils.addAll(getControllerPos(), getCorePos()),
                ArrayUtils.addAll(getAirPos(), getBasicPos()));
        return Arrays.stream(allPos).parallel()
                .map(p -> p.add(rootPos))
                .toArray(BlockPos[]::new);
    }

    public static TimeMachineStructure[] generateFromTimeMachine(ITimeMachine timeMachine) {
        String[][] layers = timeMachine.getStructureLayers();
        List<BlockPos> basicPos = new ArrayList<>();
        List<BlockPos> corePos = new ArrayList<>();
        List<BlockPos> controllerPos = new ArrayList<>();
        List<BlockPos> airPos = new ArrayList<>();

        // Decompose the string arrays in BlockPos lists
        int maxY = layers.length - 1;
        for (int y = 0; y < layers.length; y++) {
            for (int z = 0; z < layers[y].length; z++) {
                for (int x = 0; x < layers[y][z].length(); x++) {
                    BlockPos pos = new BlockPos(x, maxY - y, z);
                    switch (layers[y][z].charAt(x)) {
                        case 'Z':
                            controllerPos.add(pos);
                            break;
                        case 'C':
                            corePos.add(pos);
                            break;
                        case 'B':
                            basicPos.add(pos);
                            break;
                        case ' ':
                            airPos.add(pos);
                            break;
                        case '*':
                            break;
                        default:
                            throw new RuntimeException("Invalid Time Machine structure descriptor");
                    }
                }
            }
        }

        List<TimeMachineStructure> variants = new ArrayList<>();

        // Generate all variants of the template
        for (BlockPos rootPos: controllerPos) {
            Vec3i translation = new Vec3i(-rootPos.getX(), -rootPos.getY(), -rootPos.getZ()); // Make root be on (0, 0, 0)
            List<BlockPos> translatedBasicPos = translatePosList(new ArrayList<>(basicPos), translation);
            List<BlockPos> translatedCorePos = translatePosList(new ArrayList<>(corePos), translation);
            List<BlockPos> translatedControllerPos = translatePosList(new ArrayList<>(controllerPos), translation);
            List<BlockPos> translatedAirPos = translatePosList(new ArrayList<>(airPos), translation);

            for (BlockRotation rotation: BlockRotation.values()) { // Apply all possible rotations
                List<BlockPos> rotatedBasicPos = rotatePosList(new ArrayList<>(translatedBasicPos), rotation);
                List<BlockPos> rotatedCorePos = rotatePosList(new ArrayList<>(translatedCorePos), rotation);
                List<BlockPos> rotatedControllerPos = rotatePosList(new ArrayList<>(translatedControllerPos), rotation);
                List<BlockPos> rotatedAirPos = rotatePosList(new ArrayList<>(translatedAirPos), rotation);

                // Build the variant and add it to the return list
                TimeMachineStructure structure = new TimeMachineStructure(
                        rotatedBasicPos.toArray(new BlockPos[]{}),
                        rotatedCorePos.toArray(new BlockPos[]{}),
                        rotatedControllerPos.toArray(new BlockPos[]{}),
                        rotatedAirPos.toArray(new BlockPos[]{}),
                        timeMachine);
                variants.add(structure);
            }
        }

        return variants.toArray(new TimeMachineStructure[]{});
    }

    private static List<BlockPos> translatePosList(List<BlockPos> posList, Vec3i translation) {
        return posList.parallelStream()
                .map(p -> p.add(translation))
                .collect(Collectors.toList());
    }

    private static List<BlockPos> rotatePosList(List<BlockPos> posList, BlockRotation rotation) {
        if (rotation == BlockRotation.NONE) return posList;
        return posList.parallelStream()
                .map(p -> p.rotate(rotation))
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/com/rdvdev2/timetravelmod/impl/common/timemachine/TimeMachineStructure.kt
package com.rdvdev2.timetravelmod.impl.common.timemachine

import com.rdvdev2.timetravelmod.api.timemachine.ITimeMachine
import com.rdvdev2.timetravelmod.api.timemachine.ITimeMachineUpgrade
import com.rdvdev2.timetravelmod.api.timemachine.block.TimeMachineUpgradeBlock
import com.rdvdev2.timetravelmod.impl.ModRegistries
import net.minecraft.block.BlockState
import net.minecraft.util.BlockRotation
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Vec3i
import net.minecraft.util.registry.Registry
import net.minecraft.world.World
import java.util.*
import java.util.stream.Collectors

class TimeMachineStructure(
    private val basicPos: Array<BlockPos>,
    private val corePos: Array<BlockPos>,
    private val controllerPos: Array<BlockPos>,
    private val airPos: Array<BlockPos>,
    val timeMachine: ITimeMachine
    ) {

    private val upgradeStates: Array<BlockState> = generateUpgradeStates(timeMachine)
    private val controllerStates: Array<BlockState>
        get() = timeMachine.controllerStates
    private val coreStates: Array<BlockState>
        get() = timeMachine.coreStates
    private val basicStates: Array<BlockState>
        get() = timeMachine.basicStates

    fun checkIfBuilt(world: World, rootPos: BlockPos): Boolean {
        val basicStates = listOf(*basicStates.clone(), *upgradeStates.clone())
        val coreStates = listOf(*coreStates.clone())
        val controllerStates = listOf(*controllerStates.clone())
        return (
                Arrays.stream(controllerPos)
                    .map { it.add(rootPos) }
                    .allMatch { controllerStates.contains(world.getBlockState(it)) }
                && Arrays.stream(corePos)
                    .map { it.add(rootPos) }
                    .allMatch { coreStates.contains(world.getBlockState(it)) }
                && Arrays.stream(airPos)
                    .map { it.add(rootPos) }
                    .allMatch { world.isAir(it) }
                && Arrays.stream(basicPos)
                    .map { it.add(rootPos) }
                    .allMatch { basicStates.contains(world.getBlockState(it)) }
                )
    }

    fun getUpgrades(world: World, rootPos: BlockPos): Map<ITimeMachineUpgrade, BlockPos> {
        val upgradeStates = listOf(*upgradeStates.clone())
        return Arrays.stream(basicPos)
            .map { it.add(rootPos) }
            .filter { upgradeStates.contains(world.getBlockState(it)) }
            .collect(Collectors.toMap({ (world.getBlockState(it).block as TimeMachineUpgradeBlock).upgrade }, { it }))
    }

    fun getStructurePos(rootPos: BlockPos): Array<BlockPos> {
        val allPos = controllerPos + corePos + airPos + basicPos
        return Arrays.stream(allPos).parallel()
            .map { it.add(rootPos) }
            .toArray { arrayOfNulls<BlockPos>(it) }
    }

    companion object {

        private val UPGRADE_STATE_CACHE: MutableMap<ITimeMachine, Array<BlockState>> = HashMap()

        private fun generateUpgradeStates(timeMachine: ITimeMachine): Array<BlockState> {
            if (UPGRADE_STATE_CACHE.containsKey(timeMachine)) return UPGRADE_STATE_CACHE[timeMachine]!!
            val upgrades = ModRegistries.TIME_MACHINE_UPGRADE.stream().parallel()
                .filter { it.isTimeMachineCompatible(timeMachine) }
                .collect(Collectors.toList())
            UPGRADE_STATE_CACHE[timeMachine] = Registry.BLOCK.stream().parallel()
                .filter { it is TimeMachineUpgradeBlock && upgrades.contains(it.upgrade) }
                .flatMap { it.stateManager.states.parallelStream() }
                .toArray { arrayOfNulls<BlockState>(it) }
            return generateUpgradeStates(timeMachine)
        }

        fun generateFromTimeMachine(timeMachine: ITimeMachine): Array<TimeMachineStructure> {
            val layers = timeMachine.structureLayers
            val basicPos = mutableListOf<BlockPos>()
            val corePos = mutableListOf<BlockPos>()
            val controllerPos = mutableListOf<BlockPos>()
            val airPos = mutableListOf<BlockPos>()

            // Decompose the string arrays in BlockPos lists
            val maxY = layers.size - 1
            for (y in layers.indices) {
                for (z in 0 until layers[y].size) {
                    for (x in 0 until layers[y][z].length) {
                        val pos = BlockPos(x, maxY - y, z)
                        when (layers[y][z][x]) {
                            'Z' -> controllerPos.add(pos)
                            'C' -> corePos.add(pos)
                            'B' -> basicPos.add(pos)
                            ' ' -> airPos.add(pos)
                            '*' -> { /* IGNORE */}
                            else -> throw RuntimeException("Invalid Time Machine structure descriptor")
                        }
                    }
                }
            }
            val variants = mutableListOf<TimeMachineStructure>()

            // Generate all variants of the template
            for (rootPos in controllerPos) {
                val translation = Vec3i(-rootPos.x, -rootPos.y, -rootPos.z) // Make root be on (0, 0, 0)
                val translatedBasicPos = translatePosList(ArrayList(basicPos), translation)
                val translatedCorePos = translatePosList(ArrayList(corePos), translation)
                val translatedControllerPos = translatePosList(ArrayList(controllerPos), translation)
                val translatedAirPos = translatePosList(ArrayList(airPos), translation)
                for (rotation in BlockRotation.values()) { // Apply all possible rotations
                    val rotatedBasicPos = rotatePosList(ArrayList(translatedBasicPos), rotation)
                    val rotatedCorePos = rotatePosList(ArrayList(translatedCorePos), rotation)
                    val rotatedControllerPos = rotatePosList(ArrayList(translatedControllerPos), rotation)
                    val rotatedAirPos = rotatePosList(ArrayList(translatedAirPos), rotation)

                    // Build the variant and add it to the return list
                    val structure = TimeMachineStructure(
                        rotatedBasicPos.toTypedArray(),
                        rotatedCorePos.toTypedArray(),
                        rotatedControllerPos.toTypedArray(),
                        rotatedAirPos.toTypedArray(),
                        timeMachine
                    )
                    variants.add(structure)
                }
            }
            return variants.toTypedArray()
        }

        private fun translatePosList(posList: List<BlockPos>, translation: Vec3i): List<BlockPos> {
            return posList.parallelStream()
                .map { it.add(translation) }
                .collect(Collectors.toList())
        }

        private fun rotatePosList(posList: List<BlockPos>, rotation: BlockRotation): List<BlockPos> {
            return if (rotation == BlockRotation.NONE) posList
            else posList.parallelStream()
                .map { it.rotate(rotation) }
                .collect(Collectors.toList())
        }
    }

}